internal/impl/parquet/schema_coersion.go

```diff
@@ -0,0 +1,148 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parquet
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/gofrs/uuid/v5"
+	"github.com/parquet-go/parquet-go"
+	"github.com/parquet-go/parquet-go/format"
+)
+
+type schemaVisitor interface {
+	visitLeaf(value any, schemaNode parquet.Node) (any, error)
+}
+
+func visitWithSchema(visitor schemaVisitor, value any, schemaNode parquet.Node) (any, error) {
+	if schemaNode.Leaf() {
+		if schemaNode.Optional() && value == nil {
+			return nil, nil
+		}
+		return visitor.visitLeaf(value, schemaNode)
+	}
+
+	switch group := value.(type) {
+	case map[string]any:
+		for _, childSchemaNode := range schemaNode.Fields() {
+			name := childSchemaNode.Name()
+			if childValue, ok := group[name]; ok {
+				var err error
+				group[name], err = visitWithSchema(visitor, childValue, childSchemaNode)
+				if err != nil {
+					return nil, fmt.Errorf("visiting [%s]: %w", name, err)
+				}
+			}
+		}
+		return group, nil
+
+	case []any:
+		for i := range group {
+			var err error
+			group[i], err = visitWithSchema(visitor, group[i], schemaNode)
+			if err != nil {
+				return nil, fmt.Errorf("visiting [%d]: %w", i, err)
+			}
+		}
+		return group, nil
+
+	case nil:
+		return nil, nil
+
+	default:
+		panic(fmt.Sprintf("unexpected group value type: %T", value))
+	}
+}
+
+type encodingCoersionVisitor struct{}
+
+func (encodingCoersionVisitor) visitLeaf(value any, schemaNode parquet.Node) (any, error) {
+	logicalType := schemaNode.Type().LogicalType()
+	if logicalType == nil {
+		return value, nil
+	}
+	if logicalType.Timestamp != nil {
+		switch v := value.(type) {
+		case string:
+			ts, err := time.Parse(time.RFC3339, v)
+			if err != nil {
+				return nil, fmt.Errorf("parsing string RFC3339 timestamp: %w", err)
+			}
+			unit := logicalType.Timestamp.Unit
+			if unit.Millis != nil {
+				return ts.UnixMilli(), nil
+			} else if unit.Micros != nil {
+				return ts.UnixMicro(), nil
+			} else if unit.Nanos != nil {
+				return ts.UnixNano(), nil
+			} else {
+				panic("unreachable")
+			}
+		}
+	} else if logicalType.Json != nil {
+		switch value.(type) {
+		case map[string]any, []any:
+			jsonBytes, err := json.Marshal(value)
+			if err != nil {
+				return nil, fmt.Errorf("encoding value as JSON: %w", err)
+			}
+			return jsonBytes, nil
+		default:
+			return value, nil
+		}
+	}
+
+	return value, nil
+}
+
+type decodingCoersionVisitor struct{}
+
+func (decodingCoersionVisitor) visitLeaf(value any, schemaNode parquet.Node) (any, error) {
+	logicalType := schemaNode.Type().LogicalType()
+	if logicalType == nil {
+		return value, nil
+	}
+	if logicalType.Timestamp != nil {
+		tsNum, ok := value.(int64)
```

Comment:
https://pkg.go.dev/github.com/redpanda-data/benthos/v4@v4.44.1/public/bloblang#ValueAsTimestamp

Reply:
The fact that parquet lets the timestamp be stored as nanos, millis, or micros makes that function less useful

Reply:
What do you mean? Just because it's stored as some form doesn't mean we should interpret it differently from the bloblang semantics.

Reply:
`ValueAsTimestamp` assumes that the given `int64` is the number of seconds since the epoch. Parquet allows the timestamp to be stored as nano, milli, or microseconds, so we have to cover each case here and can't just call that function.